class java.lang.String cannot be cast to class java.lang.Boolean (java.lang.String and java.lang.Boolean are in module java.base of loader 'bootstrap')

```
Caused by: java.lang.ClassCastException: class java.lang.String cannot be cast to class java.lang.Boolean (java.lang.String and java.lang.Boolean are in
module java.base of loader 'bootstrap')
        at io.gitlab.arturbosch.detekt.api.ConfigAware$DefaultImpls.getActive(ConfigAware.kt:48)
        at io.gitlab.arturbosch.detekt.api.Rule.getActive(Rule.kt:20)
        at io.gitlab.arturbosch.detekt.api.Rule.visitCondition(Rule.kt:61)
        at io.gitlab.arturbosch.detekt.api.internal.BaseRule.visitFile(BaseRule.kt:49)
        at io.gitlab.arturbosch.detekt.core.Analyzer$analyze$4.invoke(Analyzer.kt:121)
        at io.gitlab.arturbosch.detekt.core.Analyzer.analyze(Analyzer.kt:133)
        at io.gitlab.arturbosch.detekt.core.Analyzer.runSync(Analyzer.kt:68)
        ... 147 more
```

## Expected Behavior
it's caused by miss typing of "true", but the app might show about the place of error congiguration?

## Steps to Reproduce

<!-- Provide a link to a live example, or an unambiguous set of steps to -->
<!-- reproduce this bug. Include code to reproduce, if relevant -->

## Context
<!-- How has this issue affected you? What are you trying to accomplish? -->
<!-- Providing context helps us come up with a solution that is most useful in the real world -->

## Your Environment

```
Running detekt '1.17.1' on Java '14.0.1+8' on OS 'Windows 10'
Gradle: gradle-6.9-bin.zip
Android Studio 4.2.1
Build #AI-202.7660.26.42.7351085, built on May 11, 2021
Runtime version: 11.0.8+10-b944.6842174 amd64
VM: OpenJDK 64-Bit Server VM by N/A
Windows 10 10.0
GC: G1 Young Generation, G1 Old Generation
Memory: 4096M
Cores: 24
Registry: external.system.auto.import.disabled=true
Non-Bundled Plugins: com.dubreuia, com.notime.intellijPlugin.backgroundImagePlus, org.jetbrains.kotlin, izhangzhihao.rainbow.brackets
```

- Gradle Scan Link https://gradle.com/s/qv27lp2jznzmw
- Link to your project https://github.com/tateisu/SubwayTooter

I tried to make the change, but it did not have the effect I was hoping for. The `ClassCastException` is thrown in the `ConfigAware` default implementation bypassing the catch clause. This makes sense but is rather annoying.

```
Caused by: java.lang.ClassCastException: class java.lang.String cannot be cast to class java.lang.Boolean (java.lang.String and java.lang.Boolean are in module java.base of loader 'bootstrap')
	at io.gitlab.arturbosch.detekt.api.ConfigAware$DefaultImpls.getActive(ConfigAware.kt:48)
	at io.gitlab.arturbosch.detekt.api.Rule.getActive(Rule.kt:20)
	at io.gitlab.arturbosch.detekt.api.Rule.visitCondition(Rule.kt:61)
	at io.gitlab.arturbosch.detekt.api.internal.BaseRule.visitFile(BaseRule.kt:49)
```

I thought the stack trace I wrote in the issue wasn't enough, so I tried again using cli-all.jar instead of via Gradle.

```
$ java -jar detekt-cli-1.17.1-all.jar -c config/detekt/detekt.yml
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.jetbrains.kotlin.com.intellij.util.ReflectionUtil (file:/Z:/20210630TestDetekt
/detekt-cli-1.17.1-all.jar) to method java.util.ResourceBundle.setParent(java.util.ResourceBundle)
WARNING: Please consider reporting this to the maintainers of org.jetbrains.kotlin.com.intellij.util.ReflectionUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
java.lang.IllegalStateException: Analyzing Z:\20210630TestDetekt\src\main\kotlin\Main.kt led to an exception.
The original exception message was: Value "tru" set for config parameter "empty-blocks > EmptyClassBlock > active" is no
t of required type Boolean.
Running detekt '1.17.1' on Java '14.0.1+8' on OS 'Windows 10'
If the exception message does not help, please feel free to create an issue on our GitHub page.
        at io.gitlab.arturbosch.detekt.core.AnalyzerKt.throwIllegalStateException(Analyzer.kt:152)
        at io.gitlab.arturbosch.detekt.core.AnalyzerKt.access$throwIllegalStateException(Analyzer.kt:1)
        at io.gitlab.arturbosch.detekt.core.Analyzer.runSync(Analyzer.kt:69)
        at io.gitlab.arturbosch.detekt.core.Analyzer.run(Analyzer.kt:51)
        at io.gitlab.arturbosch.detekt.core.tooling.Lifecycle$analyze$result$1.invoke(Lifecycle.kt:43)
        at io.gitlab.arturbosch.detekt.core.tooling.Lifecycle$analyze$result$1.invoke(Lifecycle.kt:22)
        at io.gitlab.arturbosch.detekt.core.util.PerformanceMonitor.measure(PerformanceMonitor.kt:42)
        at io.gitlab.arturbosch.detekt.core.tooling.Lifecycle$DefaultImpls.measure(Lifecycle.kt:30)
        at io.gitlab.arturbosch.detekt.core.tooling.Lifecycle$DefaultImpls.analyze(Lifecycle.kt:40)
        at io.gitlab.arturbosch.detekt.core.tooling.DefaultLifecycle.analyze(Lifecycle.kt:57)
        at io.gitlab.arturbosch.detekt.core.tooling.AnalysisFacade$runAnalysis$1.invoke(AnalysisFacade.kt:43)
        at io.gitlab.arturbosch.detekt.core.tooling.AnalysisFacade$runAnalysis$1.invoke(AnalysisFacade.kt:21)
        at io.gitlab.arturbosch.detekt.core.tooling.ProcessingSpecSettingsBridgeKt.withSettings(ProcessingSpecSettingsBr
idge.kt:26)
        at io.gitlab.arturbosch.detekt.core.tooling.AnalysisFacade.runAnalysis$detekt_core(AnalysisFacade.kt:42)
        at io.gitlab.arturbosch.detekt.core.tooling.AnalysisFacade.run(AnalysisFacade.kt:25)
        at io.gitlab.arturbosch.detekt.cli.runners.Runner.call(Runner.kt:33)
        at io.gitlab.arturbosch.detekt.cli.CliRunner.run(CliRunner.kt:33)
        at io.gitlab.arturbosch.detekt.cli.CliRunner.run(CliRunner.kt:15)
        at io.gitlab.arturbosch.detekt.cli.Main.main(Main.kt:19)
Caused by: java.lang.IllegalStateException: Value "tru" set for config parameter "empty-blocks > EmptyClassBlock > activ
e" is not of required type Boolean.
        at io.gitlab.arturbosch.detekt.core.config.BaseConfigKt.valueOrDefaultInternal(BaseConfig.kt:29)
        at io.gitlab.arturbosch.detekt.core.config.BaseConfigKt.valueOrDefaultInternal$default(BaseConfig.kt:15)
        at io.gitlab.arturbosch.detekt.core.config.YamlConfig.valueOrDefault(YamlConfig.kt:32)
        at io.gitlab.arturbosch.detekt.core.config.DisabledAutoCorrectConfig.valueOrDefault(DisabledAutoCorrectConfig.kt
:13)
        at io.gitlab.arturbosch.detekt.api.ConfigAware$DefaultImpls.valueOrDefault(ConfigAware.kt:65)
        at io.gitlab.arturbosch.detekt.api.Rule.valueOrDefault(Rule.kt:20)
        at io.gitlab.arturbosch.detekt.api.ConfigAware$DefaultImpls.getActive(ConfigAware.kt:48)
        at io.gitlab.arturbosch.detekt.api.Rule.getActive(Rule.kt:20)
        at io.gitlab.arturbosch.detekt.api.Rule.visitCondition(Rule.kt:61)
        at io.gitlab.arturbosch.detekt.api.MultiRule.preVisit(MultiRule.kt:19)
        at io.gitlab.arturbosch.detekt.api.internal.BaseRule.visitFile(BaseRule.kt:50)
        at io.gitlab.arturbosch.detekt.core.Analyzer$analyze$4.invoke(Analyzer.kt:121)
        at io.gitlab.arturbosch.detekt.core.Analyzer.analyze(Analyzer.kt:133)
        at io.gitlab.arturbosch.detekt.core.Analyzer.runSync(Analyzer.kt:68)
```

The output in above is the expected behavior.

There seems to be a necessary condition to output the error output mentioned in the issue.
Android Studio? No, it happened even when I incorporated the Gradle Plugin into a test project created with IntelliJ IDEA.

```
kotlin 1.5.20
detekt 1.17.1

IntelliJ IDEA 2021.1.2 (Community Edition)
Build #IC-211.7442.40, built on June 1, 2021
Runtime version: 11.0.11+9-b1341.57 amd64
VM: Dynamic Code Evolution 64-Bit Server VM by JetBrains s.r.o.
Windows 10 10.0
GC: ParNew, ConcurrentMarkSweep
Memory: 3959M
Cores: 24
Non-Bundled Plugins: com.notime.intellijPlugin.backgroundImagePlus (1.0.3), org.jetbrains.kotlin (211-1.5.20-release-284-IJ7442.40), izhangzhihao.rainbow.brackets (6.18)
Kotlin: 211-1.5.20-release-284-IJ7442.40
```

```
Caused by: java.lang.ClassCastException: class java.lang.String cannot be cast to class java.lang.Boolean (java.lang.String and java.lang.Boolean are in module java.base of loader 'bootstrap')
        at io.gitlab.arturbosch.detekt.api.ConfigAware$DefaultImpls.getActive(ConfigAware.kt:48)
        at io.gitlab.arturbosch.detekt.api.Rule.getActive(Rule.kt:20)
        at io.gitlab.arturbosch.detekt.api.Rule.visitCondition(Rule.kt:61)
        at io.gitlab.arturbosch.detekt.api.MultiRule.preVisit(MultiRule.kt:19)
        at io.gitlab.arturbosch.detekt.api.internal.BaseRule.visitFile(BaseRule.kt:50)
        at io.gitlab.arturbosch.detekt.core.Analyzer$analyze$4.invoke(Analyzer.kt:121)
        at io.gitlab.arturbosch.detekt.core.Analyzer.analyze(Analyzer.kt:133)
        at io.gitlab.arturbosch.detekt.core.Analyzer.runSync(Analyzer.kt:68)
```

--- a/detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/config/CompositeConfig.kt
+++ b/detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/config/CompositeConfig.kt
@@ -12,8 +12,12 @@ class CompositeConfig(private val lookFirst: Config, private val lookSecond: Con
     override fun subConfig(key: String): Config =
         CompositeConfig(lookFirst.subConfig(key), lookSecond.subConfig(key))
 
-    override fun <T : Any> valueOrDefault(key: String, default: T): T =
-        lookFirst.valueOrNull(key) ?: lookSecond.valueOrDefault(key, default)
+    override fun <T : Any> valueOrDefault(key: String, default: T): T {
+        if (lookFirst.valueOrNull<T>(key) != null) {
+            return lookFirst.valueOrDefault(key, default)
+        }
+        return lookSecond.valueOrDefault(key, default)
+    }
 
     override fun <T : Any> valueOrNull(key: String): T? =
         lookFirst.valueOrNull(key) ?: lookSecond.valueOrNull(key)

--- a/detekt-core/src/test/kotlin/io/gitlab/arturbosch/detekt/core/config/CompositeConfigSpec.kt
+++ b/detekt-core/src/test/kotlin/io/gitlab/arturbosch/detekt/core/config/CompositeConfigSpec.kt
@@ -46,8 +46,10 @@ class CompositeConfigSpec : Spek({
             val expectedErrorMessage = "Value \"truuu\" set for config parameter \"style > LargeClass > active\" " +
                 "is not of required type Boolean"
 
-            assertThatThrownBy { val config: Boolean = config.valueOrDefault("active", true) }
-                .isInstanceOf(IllegalStateException::class.java)
+            assertThatThrownBy {
+                val value: Boolean = config.valueOrDefault("active", true)
+                println(value)
+            }.isInstanceOf(IllegalStateException::class.java)
                 .hasMessageContaining(expectedErrorMessage)
         }
     }

--- a/detekt-core/src/test/kotlin/io/gitlab/arturbosch/detekt/core/config/CompositeConfigSpec.kt
+++ b/detekt-core/src/test/kotlin/io/gitlab/arturbosch/detekt/core/config/CompositeConfigSpec.kt
@@ -2,6 +2,7 @@ package io.gitlab.arturbosch.detekt.core.config
 
 import io.gitlab.arturbosch.detekt.test.yamlConfig
 import org.assertj.core.api.Assertions.assertThat
+import org.assertj.core.api.Assertions.assertThatThrownBy
 import org.spekframework.spek2.Spek
 import org.spekframework.spek2.style.specification.describe
 
@@ -37,6 +38,17 @@ class CompositeConfigSpec : Spek({
             val config = compositeConfig.subConfig("style").subConfig("MagicNumber")
             val value = config.valueOrDefault("ignoreNumbers", emptyList<String>())
             assertThat(value).isEqualTo(listOf("-1", "0", "1", "2", "100", "1000"))
+        }
+
+        it("should fail with a meaningful exception when boolean property is invalid") {
+            val config = compositeConfig.subConfig("style").subConfig("LargeClass")
+
+            val expectedErrorMessage = "Value \"truuu\" set for config parameter \"style > LargeClass > active\" " +
+                "is not of required type Boolean"
+
+            assertThatThrownBy { val config: Boolean = config.valueOrDefault("active", true) }
+                .isInstanceOf(IllegalStateException::class.java)
+                .hasMessageContaining(expectedErrorMessage)
         }
     }
 })

--- a/detekt-core/src/test/resources/composite-test.yml
+++ b/detekt-core/src/test/resources/composite-test.yml
@@ -7,6 +7,6 @@ style:
     ignoreHashCodeFunction: true
     ignorePropertyDeclaration: true
     ignoreAnnotation: true
-    ignoreNumbers: [ '-1', '0', '1', '2', '100', '1000' ]
+    ignoreNumbers: ['-1', '0', '1', '2', '100', '1000']
   LargeClass:
     active: truuu

--- a/detekt-core/src/test/resources/composite-test.yml
+++ b/detekt-core/src/test/resources/composite-test.yml
@@ -7,4 +7,6 @@ style:
     ignoreHashCodeFunction: true
     ignorePropertyDeclaration: true
     ignoreAnnotation: true
-    ignoreNumbers: ['-1', '0', '1', '2', '100', '1000']
+    ignoreNumbers: [ '-1', '0', '1', '2', '100', '1000' ]
+  LargeClass:
+    active: truuu

